lib/local-cluster/lib/Control/Monitoring/Server.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE RankNTypes #-}
+
+-- | How to run a monitor
+module Control.Monitoring.Server
+    ( Protocol (..)
+    , runTCPControl
+    )
+where
+
+import Prelude
+
+import Control.Monad
+    ( forever
+    , (<=<)
+    )
+import Control.Monitoring.Monitor
+    ( Monitor (..)
+    )
+import Network.Simple.TCP
+    ( HostPreference (HostAny)
+    , serve
+    )
+import Network.Socket
+    ( socketToHandle
+    )
+import System.IO
+    ( hGetLine
+    , hPutStrLn
+    )
+import Text.Read
+    ( readMaybe
+    )
+import UnliftIO
+    ( IOMode (..)
+    , MonadUnliftIO
+    , UnliftIO (..)
+    , askUnliftIO
+    , async
+    , liftIO
+    , link
+    )
+
+-- | The command you can send to the monitor
+data Protocol
+    = Pull -- ^ Pull the next trace, if in pulling state
+    | Switch -- ^ Switch between pulling and not-pulling or vice versa
+    | Observe -- ^ Observe the current state
+    | Kill -- ^ Try to kill the program on the next trace
+    deriving stock (Show, Read)
+
+-- | Run a TCP control of the monitor on a given port
+-- We use CPS even if it's not needed in this implementation as a
+-- future-proofing measure.
+runTCPControl
+    :: MonadUnliftIO m
+    => Int
+    -- ^ The port to listen on
+    -> (b -> [String])
+    -- ^ How to render the output
+    -> Monitor m a b
+    -- ^ The monitor
+    -> (() -> m c)
+    -- ^ The action to run with the tracer from the monitor
+    -> m c
+runTCPControl port renderOuptut c action = do
+    UnliftIO run <- askUnliftIO
+    liftIO
+        $ link <=< async
```

Comment:
```suggestion
```

Remove the `async` here — it's less composable. The function `serveMonitorOnTCP` will run forever, and that's fine — it's up to the caller to decide whether it wants to run anything else in parallel.

Reply:
Yes it's less composable , I will add both versions, because this has the interface I like, at least for this use case.